src/tactic/sort.lean

```diff
@@ -0,0 +1,186 @@
+inexistent terms are ignored).
+
+Finally, `sort_summands` can also be targeted to a hypothesis.  If `hp` is in the local context,
+`sort_summands [f, g] at hp` performs the rearranging at `hp`.
+
+
+##  Future work
+
+To add better heuristics, an easy approach is to change the definition of `compare_fn`.
+Right now, it simply does a case-split: we compare two expressions alphabetically, unless they are
+both `monomial <deg> <coe>ff`, in which case, we place the term last and we use `<deg>` to further
+compare.
+
+* Allow for pattern-matching in the list of terms?
+* Allow to pass `[f, -g, h, -i]` to mean that `f, h` come last and `g i` come first?
+  The `-` could be problematic if the expressions involve `neg`.
+* Improve sorting heuristic?
+* Add support for `neg` and additive groups?
+* Allow the option of not changing the given order, except for the explicitly listed terms?
+  E.g. `sort_summands only [f, g]`, extracts `f` and `g`, placing them last, but leaving the rest
+  unchanged (except for the parenthesis, which will be straightened).
+* Add optional different operations than `+`, most notably `*`?
+* Allow custom ordering?  E.g. `sort_summands using my_rel`, where `my_rel : N → N → Prop` is a
+  relation on a type (with `+`), and, if `sort_summands` encounters a sum of terms of type `N`,
+  then it will ask the user to prove the required sorting.
+ -/
+
+namespace tactic.interactive
+
+open tactic
+setup_tactic_parser
+
+/--  Takes an `expr` and returns a list of its summands. -/
+meta def get_summands : expr → list expr
+| `(%%a + %%b) := get_summands a ++ get_summands b
+| a            := [a]
+
+section with_rel
+
+/--  We sort the elements of the list `l`, making sure that the elements of `t` appear at the
+end of the list.
+
+Use as `l.sort_with_top t rel`.
+ -/
+def _root_.list.sort_with_top {N : Type*} [decidable_eq N] (l t : list N) (rel : N → N → bool) :
+  list N :=
+let tl := t.dedup.filter (∈ l) in (l.filter (∉ tl)).qsort rel ++ tl
+
+/--  Let `rel : expr → expr → bool` be a relation, `t` a list of expressions and `e` an expression.
+`sorted_sum_with_rel rel t e` returns an ordered sum of the terms of `e`, where the order is
+determined using the relation `rel`, except that the elements from the list `t` appear reversed and
+last in the sum.
+
+We use this function for expressions in an additive commutative semigroup. -/
+meta def sorted_sum_with_rel
+  (hyp : option name) (rel : expr → expr → bool) (t : list expr) (e : expr) : tactic unit :=
+match (get_summands e).sort_with_top t rel with
+| eₕ::es := do
+  e' ← es.mfoldl (λ eₗ eᵣ, mk_app `has_add.add [eₗ, eᵣ]) eₕ,
+  e_eq ← mk_app `eq [e, e'],
+  n ← get_unused_name,
+  assert n e_eq,
+  e_eq_fmt ← pp e_eq,
+  reflexivity <|>
+    `[{ simp only [add_comm, add_assoc, add_left_comm], done, }] <|>
+    -- `[{ abel, done, }] <|> -- this works too. it's more robust but also a bit slower
```

Comment:
What if we use `abel` if the user inputs `sort_summands!`? Or should we wait until we see an example of failure?

Reply:
Let's wait and see: since this tactic is only rearranging terms and changing parentheses, I think that invoking `abel` might be overkill.